shell/platform/android/io/flutter/plugin/common/BinaryMessenger.java

```diff
@@ -30,11 +30,36 @@
    * An abstraction over the threading policy used to invoke message handlers.
    *
    * <p>These are generated by calling methods like {@link
-   * BinaryMessenger#makeBackgroundTaskQueue()} and can be passed into platform channels'
-   * constructors to control the threading policy for handling platform channels' messages.
+   * BinaryMessenger#makeBackgroundTaskQueue(TaskQueueOptions)} and can be passed into platform
+   * channels' constructors to control the threading policy for handling platform channels'
+   * messages.
    */
   public interface TaskQueue {}
 
+  /**
+   * The policy for handling tasks in a TaskQueue.
+   *
+   * <p>Values: - SERIAL - Tasks will be handled serially. - CONCURRENT - Tasks can be handled in
+   * parallel.
+   */
+  public enum TaskQueueType {
+    SERIAL,
+    CONCURRENT,
+  }
+
+  /** Options that control how a TaskQueue should operate and be created. */
+  public static class TaskQueueOptions {
+    private TaskQueueType type = TaskQueueType.SERIAL;
+
+    public TaskQueueType getType() {
+      return type;
+    }
+
+    public void setType(TaskQueueType type) {
+      this.type = type;
+    }
+  }
```

Comment:
I think it is better not to provide the "setType" method, we can put type in the constructor. In addition, maybe we can provide two static field to make the developer more convenient to use. Maybe look like this:

``` java
  /** Options that control how a TaskQueue should operate and be created. */
  public static class TaskQueueOptions {

    public static final SERIAL = new TaskQueueOptions(TaskQueueType.SERIAL);

    public static final CONCURRENT = new TaskQueueOptions(TaskQueueType.CONCURRENT);

    private final TaskQueueType type;

    public TaskQueueOptions(TaskQueueType type) {
      this.type = type;
    }

    public TaskQueueType getType() {
      return type;
    }
  }
```

Reply:
The reason I didn't do this is because it doesn't scale with the addition of new parameters.  I generally don't like mutable objects but this is essentially the Builder design pattern, where this object's only job is to represent how to create something.  In the future we may add another field like "priority".  If the object is immutable we'll have to add "priority" to the constructor and that would be a breaking change, unless we create a new constructor.  That doesn't scale well were each new field needs a new constructor and the order of arguments is arbitrarily based on the order they were added.

Same thing with the static fields, what happens when we add "priority"?  [SERIAL_HIGH, SERIAL_LOW, CONCURRENT_HIGH, CONCURRENT_LOW]?  The static fields help avoiding allocations, but clients can do that from their side instead of putting it in the API.

Reply:
I think the more Java-y thing to do here would be to make a builder class though right? So rather than providing a mutable options object to the end class, you have a builder class that knows how to provide all the right parameters to the end class. E.g. what we have for [FlutterInjector.Builder](https://github.com/flutter/engine/blob/master/shell/platform/android/io/flutter/FlutterInjector.java#L118)

Reply:
@dnfield are you suggesting:
```
taskQueue =
  binaryMessenger.getBackgroundTaskQueueBuilder().setPriority(100).setType(SERIAL).build();
```
vs
```
taskQueue =
  binaryMessenger.makeBackgroundTaskQueue((new TaskQueueOptions()).setType(SERIAL).setPriority(100)));
```

I don't think that is really more java-y.  It isn't as discoverable as having `makeBackgroundTaskQueue` overloaded.  The builder will have setters just like a builder will have.

(I added `return this` at the end of the setter, to match the pattern in FlutterInjector.Builder)

Reply:
I'm not a java style person.  Maybe just call it an options builder?

Reply:
I don't think options requires a builder since it is effectively operating like a builder.  That would be an extra layer of indirection.  Even calling it a builder would just be confusing.  java.net.SocketOptions follows the same pattern as this, it's called options and has set get methods on it, and is used as arguments.

Reply:
It makes sense to me.